File: core/transcription/src/commonMain/kotlin/dev/audiocompanion/transcription/PcmWav.kt
package dev.audiocompanion.transcription

object PcmWav {
    private const val HEADER_BYTES = 44
    private const val PCM_FORMAT = 1

    fun encodeMono16(pcm: ByteArray, sampleRateHz: Int): ByteArray {
        val out = headerMono16(pcm.size, sampleRateHz) + pcm
        return out
    }

    fun headerMono16(pcmSizeBytes: Int, sampleRateHz: Int): ByteArray {
        require(sampleRateHz > 0) { "sampleRateHz must be positive" }
        require(pcmSizeBytes >= 0) { "pcmSizeBytes must be non-negative" }
        val out = ByteArray(HEADER_BYTES)
        out.writeAscii(0, "RIFF")
        out.writeU32Le(4, (36 + pcmSizeBytes).toUInt())
        out.writeAscii(8, "WAVE")
        out.writeAscii(12, "fmt ")
        out.writeU32Le(16, 16u)
        out.writeU16Le(20, PCM_FORMAT.toUShort())
        out.writeU16Le(22, 1u)
        out.writeU32Le(24, sampleRateHz.toUInt())
        out.writeU32Le(28, (sampleRateHz * Short.SIZE_BYTES).toUInt())
        out.writeU16Le(32, Short.SIZE_BYTES.toUShort())
        out.writeU16Le(34, 16u)
        out.writeAscii(36, "data")
        out.writeU32Le(40, pcmSizeBytes.toUInt())
        return out
    }

    private fun ByteArray.writeAscii(offset: Int, value: String) {
        value.encodeToByteArray().copyInto(this, offset)
    }

    private fun ByteArray.writeU16Le(offset: Int, value: UShort) {
        this[offset] = (value.toInt() and 0xFF).toByte()
        this[offset + 1] = ((value.toInt() ushr 8) and 0xFF).toByte()
    }

    private fun ByteArray.writeU32Le(offset: Int, value: UInt) {
        this[offset] = (value.toInt() and 0xFF).toByte()
        this[offset + 1] = ((value.toInt() ushr 8) and 0xFF).toByte()
        this[offset + 2] = ((value.toInt() ushr 16) and 0xFF).toByte()
        this[offset + 3] = ((value.toInt() ushr 24) and 0xFF).toByte()
    }
}

File: ios/PebbleAudioKit/Sources/AudioCodec/PcmWav.swift
import Foundation

/// Standard 44-byte RIFF/WAVE wrapping for s16le mono PCM. Port of the KMP `PcmWav` object.
public enum PcmWav {
    public static let headerBytes = 44
    private static let pcmFormat: UInt16 = 1

    public static func encodeMono16(pcm: Data, sampleRateHz: Int) -> Data {
        var out = headerMono16(pcmSizeBytes: pcm.count, sampleRateHz: sampleRateHz)
        out.append(pcm)
        return out
    }

    public static func headerMono16(pcmSizeBytes: Int, sampleRateHz: Int) -> Data {
        precondition(sampleRateHz > 0, "sampleRateHz must be positive")
        precondition(pcmSizeBytes >= 0, "pcmSizeBytes must be non-negative")
        var out = [UInt8](repeating: 0, count: headerBytes)
        writeAscii(&out, at: 0, "RIFF")
        writeU32Le(&out, at: 4, UInt32(truncatingIfNeeded: 36 + pcmSizeBytes))
        writeAscii(&out, at: 8, "WAVE")
        writeAscii(&out, at: 12, "fmt ")
        writeU32Le(&out, at: 16, 16)
        writeU16Le(&out, at: 20, pcmFormat)
        writeU16Le(&out, at: 22, 1) // channels
        writeU32Le(&out, at: 24, UInt32(truncatingIfNeeded: sampleRateHz))
        writeU32Le(&out, at: 28, UInt32(truncatingIfNeeded: sampleRateHz * 2)) // byte rate
        writeU16Le(&out, at: 32, 2) // block align
        writeU16Le(&out, at: 34, 16) // bits per sample
        writeAscii(&out, at: 36, "data")
        writeU32Le(&out, at: 40, UInt32(truncatingIfNeeded: pcmSizeBytes))
        return Data(out)
    }

    private static func writeAscii(_ out: inout [UInt8], at offset: Int, _ value: String) {
        for (i, byte) in value.utf8.enumerated() {
            out[offset + i] = byte
        }
    }

    private static func writeU16Le(_ out: inout [UInt8], at offset: Int, _ value: UInt16) {
        out[offset] = UInt8(value & 0xFF)
        out[offset + 1] = UInt8((value >> 8) & 0xFF)
    }

    private static func writeU32Le(_ out: inout [UInt8], at offset: Int, _ value: UInt32) {
        out[offset] = UInt8(value & 0xFF)
        out[offset + 1] = UInt8((value >> 8) & 0xFF)
        out[offset + 2] = UInt8((value >> 16) & 0xFF)
        out[offset + 3] = UInt8((value >> 24) & 0xFF)
    }
}
